NetKet/Graph/lattice_impl.hpp

```diff
@@ -0,0 +1,430 @@
+
+  return distance;
+};
+std::vector<std::vector<int>> Lattice::LatticeNeighbours(int iatom) const {
+  std::vector<std::vector<int>> neighbours_matrix_in;
+  std::vector<std::vector<int>> result;
+  std::vector<double> distance;
+  double min_distance;
+
+  neighbours_matrix_in = PossibleLatticeNeighbours();
+
+  distance = NeighboursSquaredDistance(neighbours_matrix_in, iatom);
+
+  min_distance = *min_nonzero_elem(distance.begin(), distance.end());
+
+  for (std::size_t i = 0; i < distance.size(); i++) {
+    if (RelativelyEqual(distance[i], min_distance, 1.0e-6)) {
+      std::vector<int> temp;
+      temp = neighbours_matrix_in[i / natoms_];
+      temp.push_back(i % natoms_);
+      result.push_back(temp);
+    }
+  }
+
+  return result;
+}
+
+std::vector<int> Lattice::FindNeighbours(int k, int iatom) const {
+  std::vector<int> result;
+  std::vector<int> n = Site2Vector(k);
+  std::vector<int> single_neighbour;
+
+  std::vector<std::vector<int>> lattice_neighbours_vector =
+      LatticeNeighbours(iatom);
+
+  for (std::size_t i = 0; i < lattice_neighbours_vector.size(); i++) {
+    std::vector<int> single_neighbour_vector;
+
+    bool flag = true;
+    for (int j = 0; j < Ndim(); j++) {
+      int new_index = n[j] + lattice_neighbours_vector[i][j];
+      if (pbc_[j]) {
+        if (new_index < 0)
+          new_index = extent_[j] - 1;
+        else if (new_index >= extent_[j])
+          new_index = 0;
+        single_neighbour_vector.push_back(new_index);
+
+      } else if (new_index >= 0 and new_index < extent_[j])
+        single_neighbour_vector.push_back(new_index);
+
+      else
+        flag = false;  // if new_index refers to a border site, don't add this
+                       // site to the neighbours
+    }
+    if (flag) {
+      single_neighbour.push_back(Vector2Site(single_neighbour_vector) +
+                                 lattice_neighbours_vector[i][Ndim()] *
+                                     nlatticesites_);
+    }
+  }
+
+  return single_neighbour;
+}
+
+// Edges and AdjacencyList
+std::vector<Lattice::Edge> Lattice::BuildEdges() const {
+  std::vector<Lattice::Edge> edge_vector;
+  for (int k = 0; k < nlatticesites_; k++) {
+    for (int iatom = 0; iatom < natoms_; iatom++) {
+      std::vector<int> neighbours;
+
+      neighbours = Lattice::FindNeighbours(k, iatom);
+
+      for (std::size_t i = 0; i < neighbours.size(); i++) {
+        if (k + iatom * nlatticesites_ < neighbours[i])
+          edge_vector.push_back({{k + iatom * nlatticesites_, neighbours[i]}});
+      }
+    }
+  }
+  return edge_vector;
+};
+
+std::vector<std::vector<int>> Lattice::AdjacencyList() const noexcept {
+  std::vector<std::vector<int>> adlist(nsites_);
+  for (int k = 0; k < nsites_; k++) {
+    for (std::size_t i = 0; i < edges_.size(); i++) {
+      if (edges_[i][0] == k)
+        adlist[k].push_back(edges_[i][1]);
+      else if (edges_[i][1] == k)
+        adlist[k].push_back(edges_[i][0]);
+    }
+  }
+  return adlist;

```

Comment:
There is a function for that already in `netket::detail`: `AdjacencyListFromEdges`